packages/flutter_test/lib/src/widget_tester.dart

```diff
@@ -2,198 +2,436 @@
-  /// acknowledged with a call to this method, then when the second exception is
-  /// thrown, they are both dumped to the console and then the second is
-  /// rethrown from the exception handler. This will likely result in the
-  /// framework entering a highly unstable state and everything collapsing.
+  /// Use it to travel into the future without actually waiting, control the
+  /// flushing of microtasks and timers.
+  FakeAsync get async => elementTreeTester.async;
+
+  /// Runs all remaining microtasks, including those scheduled as a result of
+  /// running them, until there are no more microtasks scheduled.
   ///
-  /// It's safe to call this when there's no pending exception; it will return
-  /// null in that case.
-  dynamic takeException() {
-    dynamic result = _pendingException;
-    _pendingException = null;
-    return result;
+  /// Does not run timers. May result in an infinite loop or run out of memory
+  /// if microtasks continue to recursively schedule new microtasks.
+  void flushMicrotasks() {
+    elementTreeTester.async.flushMicrotasks();
+  }
+
+  /// Checks if the element identified by [finder] exists in the tree.
+  bool exists(Finder finder) => finder.find(this).isNotEmpty;
+
+  /// All widgets currently live on the UI returned in a depth-first traversal
+  /// order.
+  Iterable<Widget> get widgets {
+    return this.elementTreeTester.allElements
+      .map((Element element) => element.widget);
+  }
+
+  /// Finds the first widget, searching in the depth-first traversal order.
+  Widget widget(Finder finder) {
+    return finder.findFirst(this).widget;
+  }
+
+  /// Finds the first state object, searching in the depth-first traversal order.
+  State stateOf(Finder finder) {
+    Element element = finder.findFirst(this);
+    Widget widget = element.widget;
+
+    if (widget is StatefulWidget)
+      return (element as StatefulElement).state;
+
+    throw new ElementNotFoundError(
+      'Widget of type ${widget.runtimeType} found by ${finder.description} does not correspond to a StatefulWidget'
+    );
+  }
+
+  /// Emulates a tapping action at the center of the widget found by [finder].
+  void tap(Finder finder, { int pointer: 1 }) {
+    tapAt(getCenter(finder), pointer: pointer);
+  }
+
+  /// Emulates a tapping action at the given [location].
+  ///
+  /// See [ElementTreeTester.tapAt] for details.
+  void tapAt(Point location, { int pointer: 1 }) {
+    elementTreeTester.tapAt(location, pointer: pointer);
+  }
+
+  /// Scrolls by dragging the center of a widget found by [finder] by [offset].
+  void scroll(Finder finder, Offset offset, { int pointer: 1 }) {
+    scrollAt(getCenter(finder), offset, pointer: pointer);
+  }
+
+  /// Scrolls by dragging the screen at [startLocation] by [offset].
+  ///
+  /// See [ElementTreeTester.scrollAt] for details.
+  void scrollAt(Point startLocation, Offset offset, { int pointer: 1 }) {
+    elementTreeTester.scrollAt(startLocation, offset, pointer: pointer);
+  }
```

Comment:
We should really rename `scroll` and `scrollAt` to `drag` and `dragAt`, but that's unrelated to the rest of your change.